### ISSUE TITLE
IfcMaterialLayer Priority : Enforce range [0...100] ?

### ISSUE THREAD

**Issue by Gorgious56** [IfcOpenShell/IfcOpenShell#2545]
According to the docs, unless I'm mistaken the "Priority" attribute value of [IfcMaterialLayer](https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/schema/ifcmaterialresource/lexical/ifcmateriallayer.htm) should be in the enclosed range [0...100].

There is currently nothing to prevent users to set the attribute outside of this bound. I don't know if :
- This shouldn't be enforced anywhere (in this case this issue can be closed)
- This shouldn't be enforced but we add a warning popup next to the field to warn users the value is not valid
- This should be enforced by a callback to the field in Blenderbim after the user validates the value
![image](https://user-images.githubusercontent.com/25156105/198582224-8d82e255-1222-496e-a65d-b5a54905c6b5.png)
- This should be enforced in Blenderbim when the user validates the material attributes editing state
![image](https://user-images.githubusercontent.com/25156105/198582482-a1ed13bd-27d8-47b3-8e93-28513eb0f33e.png)
- This should be enforced by the ifcopenshell api when the validation occurs. In this case maybe values < 0 should be clamped to 0 and values > 100 should be clamped to 100 ?

It could be part of a broader project for other attributes that may need some kind of state validation. I would be happy to contribute if need be.

**Comment by Moult** [comment]
There aren't too many of these. I think the easiest thing to do is to do it in the API. Would you like to have a shot?

**Comment by Gorgious56** [comment]
Sure, I'll dig around but if you have some pointers for where to begin I'm all ears :)

**Comment by Moult** [comment]
Just clamp it here I guess:

 - https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/material/add_layer.py
 - https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/material/edit_layer.py

**Comment by aothms** [comment]
How about this: we build an IDS spec for these kind of "soft" constraints in the schema. That way users can exercise some control over which of these constraints they want to honor. As a first immediate result it would allow us to do validation on the in-memory or exported IFC model. In a later stage we would be able to infer the correct blenderbim controls based on the IDS. All in a vendor neutral way; the entire community can benefit.

I can think of IfcShapeRepresentation.RepresentationIdentifier and IfcMaterial.Category, but there are probably many many more.

**Comment by Moult** [comment]
It seems we can do some getter/setters to also clamp in the UI: https://blender.stackexchange.com/questions/18686/setting-a-propertys-min-and-max-values-after-creation

**Comment by Gorgious56** [comment]
Yup we would have to add custom rules for `int_value` and `float_value`. 

I don't think there is a way to automatically fetch these rules though ? Anyways if they're few and far between it can be hardcoded somewhere I guess

**Comment by Moult** [comment]
They're relatively rare, which is why I haven't bothered so far.

**Comment by Gorgious56** [comment]
So this issue was solved as far as I'm concerned but I think there is still a design discussion for how / if it should be enforced at the API level or at the interface level, or both. And if it should be a soft limit and not destructively enforced to the user. 

I don't have enough experience with the ifc specs to know if there is a case where a user wouldn't want to follow the constraints and where it would be a problem that they can't input a value outside of the constraints sugested in the schema.

I would also be nice to have a library of such constraints on attributes where applicable.

**Comment by Moult** [comment]
This is uncharted territory, so whether it should be a soft or hard limit is for us to decide :) I guess we just wait and see for more people to use it.

**Comment by BIMvoice** [comment]
Verified fixed: the Priority [0,100] clamp from #2562 is still active in current code (`ATTRIBUTE_MIN_MAX_CONSTRAINTS` in `bim/helper.py`, survived the blenderbim to bonsai rename). Reporter already confirmed this resolved their concern back then. Please reopen if the field-level clamp isn't working for you.
